src/segment-loader.js

```diff
@@ -1178,6 +1221,187 @@ export default class SegmentLoader extends videojs.EventTarget {
+      let start = 0;
+      let end = 0;
+      let sliceAppend = () => {
+        if (slices.length) {
+          sourceBuffer.appendBuffer(slices.shift());
+        } else {
+          sourceBuffer.removeEventListener('updateend', sliceAppend);
+          callback();
+        }
+      };
+
+      while (start < bytesTried) {
+        end += 5000000;
+        if (end > bytesTried) {
+          end = bytesTried;
+        }
+        slices.push(segmentBytes.slice(start, end));
+        start += 5000000;
+      }
+      sourceBuffer.addEventListener('updateend', sliceAppend);
+      sourceBuffer.appendBuffer(slices.shift());
+      return;
+    }
+
+    let bytesTrimmable;
+    let removeToTime;
+    let timeupdateHandler;
+    let tryAppend = () => {
+      if (!pendingAppend) {
+        return;
+      }
+      this.hls_.tech_.off('timeupdate', timeupdateHandler);
+      this.logger_('Appended deferred segment with ' +
+        (this.buffered_().end(0) - this.currentTime_()) + ' to spare...');
+      // we don't want this to be called again
+      sourceBuffer.removeEventListener('updateend', tryAppend);
+      try {
+        sourceBuffer.appendBuffer(segmentBytes);
+        callback();
+        pendingAppend = false;
+      } catch (error) {
+        videojs.error('QuoteExceededError');
+      }
+    };
+
+    sourceBuffer.addEventListener('updateend', tryAppend);
+
+    if (safeRemovePoint) {
+      removeToTime = safeRemovePoint;
+    } else {
+      removeToTime = safeBackBufferTrimTime(this.seekable_(),
+                                              this.currentTime_(),
+                                              this.playlist_.targetDuration || 10,
+                                              this.backBufferLength_());
+    }
+
+    bytesTrimmable = this.minBufferedBytes(0, removeToTime);
+
+    if (bytesTrimmable >= bytesTried) {
+      this.removeWholeSegments(removeToTime);
+    } else {
+      let waitUntil = this.currentTime_() + this.playlist_.targetDuration;
+      let buffered = this.buffered_();
+      let dangerZone = buffered.end(buffered.length - 1) - 2;
+
+      timeupdateHandler = () => {
+        let currentTime = this.currentTime_();
+
+        if (currentTime < waitUntil) {
+          return;
+        } else if (currentTime > dangerZone) {
+          this.removeWholeSegments(currentTime - this.playlist_.targetDuration);
+          return;
+        }
+        removeToTime = safeBackBufferTrimTime(this.seekable_(),
+                                                currentTime,
+                                                this.playlist_.targetDuration || 10,
+                                                this.backBufferLength_());
+        bytesTrimmable = this.minBufferedBytes(0, removeToTime);
+        if (bytesTrimmable >= bytesTried) {
+          this.removeWholeSegments(removeToTime);
+        }
+      };
+      this.hls_.tech_.on('timeupdate', timeupdateHandler);
+      this.on('reseteverything', () => {
```

Comment:
some of this should already be happening in the `resetEverything` method, I would advise moving anything that explicitly needs to be reset to there

Reply:
Can't, unless you're suggesting I also make `pendingAppend` and `timeupdateHandler` properties of `SegmentLoader`

Reply:
I mean, just follow through the code. You'll see that `this.state` is `READY` whenever it's idle / ready to do things. Of particular importance:
- [Don't download new segments if our state isn't `READY`](https://github.com/squarebracket/videojs-contrib-hls/blob/79f53fed214a241df7e02801fb52deb1ca662ce1/src/segment-loader.js#L581-L583)
- [Set state to `APPENDING` when we have data to append to a buffer](https://github.com/squarebracket/videojs-contrib-hls/blob/79f53fed214a241df7e02801fb52deb1ca662ce1/src/segment-loader.js#L1196)
- [Set the state to `READY` on `updateend`](https://github.com/squarebracket/videojs-contrib-hls/blob/79f53fed214a241df7e02801fb52deb1ca662ce1/src/segment-loader.js#L1441)

If you follow through all of them you'll see that the state is reset to `READY` in a number of circumstances.

If we didn't reset to `READY`, a seek would do nothing because no segments would be downloaded and it would just be stuck forever with no buffer.

Reply:
The state should be set to ready by the call to [abort](https://github.com/videojs/videojs-contrib-hls/blob/1eaea13c1148e14528176ca4275e8fb3652c5687/src/segment-loader.js#L231) which is triggered from any calls to `resetEverything()`. Perhaps there is something else blocking the state change back to READY from occurring? It's worth revisiting these changes when updating the other portions of this PR to ensure we are only doing what is necessary in this handler.

Once source-updater.js is updated to use a new variable instead of nulling out pendingCallback_, I'm fairly sure you won't need `pendingAppend` anymore? Would be good to double check again.